### ISSUE TITLE
[$1000] Getting errors on other workspaces too if you get an error in a single workspace

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#16830]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Go to staging dot on chrome web
2. Create a new workspace
3. Change the name of the workspace and type in multiple emojis until you get an error
4. Click on save
5. Now go to other workspaces general settings and see that each of the workspace also displays the same error . An error in a single workspace is affecting other workspaces too. Also it's affecting when a new workspace is created.

## Expected Result: 
Other workspaces should not be affected if an error arises in a single workspace
## Actual Result: 
Other workspaces also gets the same error when one workspaces gets the error on general settings page

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.93-4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/229234513-20d40c37-8416-4365-a5ef-f79ed5d177ea.mp4

https://user-images.githubusercontent.com/43996225/229234528-7bb5bc8b-8324-409f-98d4-52723bdfbe1b.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @priya-zha
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1680274078505429

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01e5262531be62da28</li>
        <li>Upwork Job ID: 1642845457274187776</li>
        <li>Last Price Increase: 2023-04-03</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @laurenreidexpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @sobitneupane (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @AndrewGable (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by laurenreidexpensify** [comment]
Sorry took me a while longer than normal to review, can confirm all checks out and ready for external

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Workspace error is "shared" to all workspaces.

### What is the root cause of that problem?
The error comes from server response in this form of json
<img width="272" alt="image" src="https://user-images.githubusercontent.com/50919443/229493948-0554059f-f97a-4d4d-92ea-886112995812.png">

Focus on the `onyxData`. We can see that the key of the onyx is `workspaceSettingsForm` which is also the key of the workspace setting `Form` component.

https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/pages/workspace/WorkspaceSettingsPage.js#L89-L90

This means, all workspace have the same key, so when we receive an error on one workspace, all the workspace will also show the same error.

### What changes do you think we should make in order to solve the problem?


1. First, the response should be fixed from the BE by returning the `workspaceSettingsForm` appended with the workspace/policy id (i.e., `workspaceSettingsForm_123`)
2. On the FE, we should update the `Form` key by appending it with the workspace/policy id (the same thing as above).

**Comment by tienifr** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When typing in multiple emojis into the workspace name, we get a back-end invalid length error, and the error propagates to other workspaces.

### What is the root cause of that problem?
There're 2 problems here:
1. The emojis-filled workspace name escapes the max length validation in the front-end side and was sent to the back-end, leading to errors. This doesn't happen with no-emoji workspace name. This leads to the user able to save the invalid name as optimistic data and the wrong name shows up all over the app, we should make sure to validate it correctly.
This is because in the back-end, we're using C's `strlen` to calculate the length of the string.
So `😁` is 2 in client side, 4 in server side.
`😶‍🌫️` is 6 in client side, 14 in server side.
`a` is 1 both in client and server side.
We've had this issue before with the comment length [here](https://github.com/Expensify/App/issues/13988).

2. The error propagates to other workspaces
This is because the `formId` is being set to the same string for all workspaces here https://github.com/Expensify/App/blob/2123e1e7788e2aeeca5782c9a150ce4de3f386d1/src/pages/workspace/WorkspaceSettingsPage.js#L90, so all workspaces share the same error.

### What changes do you think we should make in order to solve the problem?

1. To fix this we need to add a client-side validation so that the length is calculated consistently with the back-end, and there'll be no error thrown.

Since the C's `strlen` is calculating the byte of the string, we can do the same in JS by using the `TextEncoder` and get the `byteLength` property of the text

```
const getStringLength = (text) => {
   const encoder = new TextEncoder();
   const encodedText = encoder.encode(text);
   return encodedText.byteLength;   
}
```

The validation can be added to this function https://github.com/Expensify/App/blob/6369026a0aa83766ea4f72c759f3ff9281a00750/src/pages/workspace/WorkspaceSettingsPage.js#L65

2. To fix this we need to make the `formId` unique for each workspace, we can use `${ONYXKEYS.FORMS.WORKSPACE_SETTINGS_FORM}${this.props.policy.id}` as the `formId` in client side, and the back-end will need to return the form id in that format as well.

### What alternative solutions did you explore? (Optional)
- To fix `1` above, another way is to change the BE to calculate the string length in similar way to the FE, but as mentioned in [a similar issue](https://github.com/Expensify/App/issues/13988), changing the back-end is expensive so we opted for changing the front-end, so I assume we might want the same here. 
- We'll want to fix the `2` for all forms with similar issues as well, currently we're using a fixed string for many such forms that share the same UI.

**Comment by bernhardoj** [comment]
That won't solve the initial issue of this GH, right?. If there is another server error, the "shared error" issue will happen again. I think the different length validation should be handled on a separate issue.

**Comment by tienifr** [comment]
@bernhardoj thanks for the concern, the "shared error" will be solved with the 2nd part of [my proposal](https://github.com/Expensify/App/issues/16830#issuecomment-1494318854), that part is similar to your proposal.

> I think the different length validation should be handled on a separate issue.

The different length issue is what leads the user to notice this bug so I think it's part of the issue and can be handled here.

**Comment by bernhardoj** [comment]
It's related, but has a different root cause. We previously have a case where even we have the same root cause, yet handled on a separate issue https://github.com/Expensify/App/issues/15925. Why this case would be different? There is also another related issue that @hoangzinh mentioned on his proposal. Why we ignore that? We can simply revert the old workspace value in `failureData`. I'm ignoring the related issues as it is not the scope of this GH.

I would love to hear and accept any decision from @AndrewGable.

**Comment by tienifr** [comment]
> There is also another related issue that @hoangzinh mentioned on his proposal. Why we ignore that? We can simply revert the old workspace value in `failureData`

@bernhardoj IMO that's expected, not an issue. If we revert the old workspace value the user will have no idea what happened (let's say if the user updates the workspace name when offline, then comes back online).

Anyway that case will not happen if we have proper validation from client-side.

**Comment by laurenreidexpensify** [comment]
@sobitneupane @AndrewGable bump on review ^^

**Comment by sobitneupane** [comment]
https://github.com/Expensify/App/issues/16830#issuecomment-1495540185

cc: @AndrewGable

**Comment by MelvinBot** [comment]
📣 @tienifr You have been assigned to this job by @AndrewGable!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by AndrewGable** [comment]
Sorry for delay, agree with @sobitneupane. Assigned @tienifr.

**Comment by tienifr** [comment]
As mentioned in my [proposal](https://github.com/Expensify/App/issues/16830#issuecomment-1494318854):

> To fix this we need to make the formId unique for each workspace, we can use ${ONYXKEYS.FORMS.WORKSPACE_SETTINGS_FORM}${this.props.policy.id} as the formId in client side, and the back-end will need to return the form id in that format as well.

I believe we need to put this issue on hold until BE change is made. Cc: @laurenreidexpensify

**Comment by tienifr** [comment]
@laurenreidexpensify Just a friendly remind in case you missed this!

**Comment by laurenreidexpensify** [comment]
thanks @tienifr - @AndrewGable are we good to hold this one?

**Comment by laurenreidexpensify** [comment]
Note to myself: I have sent offers to C, C+ and bug reporter in Upwork, but still waiting on @AndrewGable confirmation re: hold

**Comment by tienifr** [comment]
Bump @AndrewGable in case you missed it: https://github.com/Expensify/App/issues/16830#issuecomment-1501361798.

In addition, while we're waiting for BE update, I think we should proceed with reviewing the PR: https://github.com/Expensify/App/pull/16958. @sobitneupane What do you think?

**Comment by sobitneupane** [comment]
@tienifr As per my understanding the change in backend is required to prevent such issues in the future. For this specific case, your PR should be sufficient to solve the issue, right?

**Comment by tienifr** [comment]
My proposal has two parts:
1. Client-side validation
2. `formID` update (require BE change as well)

Therefore, **my PR would fully work if BE was updated**. Otherwise, only the first one is resolved.

For example, if it were another issue not related to workspace name and causing BE error, user wouldn't get notified; since BE's `formID` had not been updated to match client's `formID` as in my PR.

**Comment by sobitneupane** [comment]
In that case, I will prefer reviewing the PR only after the change in backend. Overall, PR looks good to me.

**Comment by MelvinBot** [comment]
@AndrewGable, @sobitneupane, @laurenreidexpensify, @tienifr Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by laurenreidexpensify** [comment]
@AndrewGable do we have an idea of where the change in the backend is being made? I'm a bit confused here if we're still held on a specific issue, or whether we need to write another PR for the change referred to here?

**Comment by laurenreidexpensify** [comment]
Starting a convo in Slack https://expensify.slack.com/archives/C049HHMV9SM/p1680274078505429

**Comment by laurenreidexpensify** [comment]
Still discussing backend change in Slack in thread above

**Comment by laurenreidexpensify** [comment]
Conversation ongoing in Slack

**Comment by laurenreidexpensify** [comment]
Bumped Slack thread again

**Comment by laurenreidexpensify** [comment]
Bumped Slack thread again

**Comment by laurenreidexpensify** [comment]
As discussed in Slack today, we can no longer reproduce this so am closing. thank you

**Comment by laurenreidexpensify** [comment]
Folks I will review the comments here tomorrow and give a recommendation. thanks for the flag!
